### PR TITLE
[guilib] Add animated image/texture support via ffmpeg.

### PR DESCRIPTION
**Pull request by ace20022**
**Warning:** In its current state, this pr can cause memleaks and out of memory crashes.

### CODE REVIEW

**Comment by ace20022** [comment]
@garbear @notspiff fyi. The same as for gifs applies here: As long as we don't load them in a job and each/some frames on demand, i's really only usable for small apngs.

@fritsch It really is a quick "hack". Are you in the mood to have a look?

**Comment by fritsch** [comment]
Hi @ace20022 thx much. Currently ill and not that fit. Can check over the weekend. Android keeps me busy currently.

**Comment by ace20022** [comment]
@fritsch np. Get well soon!
Nothing urgent, I was just happy that it works (more or less) ;)

**Comment by ace20022** [comment]
:cry: libgif can die now :cry:

**Comment by mgonzales71** [comment]
personally want to thank you guys for finally bringing ANPG support to Kodi.  Willing to test this if you create a test build - either way look forward to enjoying this at some future point in Kodi.  Many thanks again for all your hard work! :smile:

**Comment by fritsch** [comment]
Which platform do you need @mgonzales71 ?

@ace20022 i have not forgotten to review but still busy with AudioTrack - sorry for that!

**Comment by mgonzales71** [comment]
@fritsch using Kodi on Win 10 x64 Pro mainly here, also occassionally on Android (ARM) so can test both if needed.

**Comment by ace20022** [comment]
np, I'm still hunting a mem leak, hoping that it's not in ffmepg's png decoder (Valgrind indicates that)

@fritsch @mgonzales71 don't use a build of the pr in this state! A build is only useful without the two leak check commits.

**Comment by mgonzales71** [comment]
@ace20022 I am fine with waiting once you squash mem leak(s) 

FYI - a great collection of APNG resource info - http://littlesvr.ca/apng/ for the interested

**Comment by ace20022** [comment]
@fritsch thx for the review, found two leaks, an old one and a new one ;)
I've not implemented a memory restriction...
@MilhouseVH could you please test what happens on a pi with the images from: https://www.reddit.com/r/apng ?
Note the files must have the extension apng (not png).

jenkins build this please

**Comment by MilhouseVH** [comment]
> could you please test what happens on a pi with the images from: https://www.reddit.com/r/apng ?

On Pi (RPi2) and also x86, with this PR the pngs (`*.apng`) will be listed in Pictures - without the PR they wouldn't even be listed - and they will display, but there is no animation, presumably just the first frame is being displayed.

Were you expecting animation? It could be an OpenELEC issue, although I added `--enable-encoder=apng` to ffmpeg, but it had no effect.

Looking in the texture cache, the cached version of the 20MB+ `apng` files is being created as 400KB `.png` files, which might be right - I doubt we'd want the `apng` files to hit the cache, would we?

**Comment by fritsch** [comment]
@ace20022 could you show me the leaks that you fixed?

File: xbmc/guilib/FFmpegImage.cpp [hunk 1]
```diff
@@ -395,6 +474,7 @@ bool CFFmpegImage::Decode(unsigned char * const pixels, unsigned int width, unsi
   }
   pictureRGB->data[0] = nullptr;
   avpicture_free(pictureRGB);
+  av_free(pictureRGB);
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Mini leak

File: xbmc/guilib/FFmpegImage.cpp [hunk 2]
```diff
@@ -188,79 +238,103 @@ bool CFFmpegImage::LoadImageFromMemory(unsigned char* buffer, unsigned int bufSi
...
+  if (!m_fctx)
+  {
+    return nullptr;
+  }
+
+  AVStream *st = m_fctx->streams[0];
+  if (!st)
+  {
+    return nullptr;
+  }
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Reminder: merge checks

File: xbmc/guilib/FFmpegImage.cpp [hunk 3]
```diff
@@ -124,44 +146,70 @@ CFFmpegImage::~CFFmpegImage()
   av_frame_free(&m_pFrame);
   // someone could have forgotten to call us
   CleanupLocalOutputBuffer();
+  if (m_ioctx)
+    FreeIOCtx(&m_ioctx);
+  if (m_fctx)
+  {
+    for (unsigned int i = 0; i < m_fctx->nb_streams; i++) {
+      avcodec_close(m_fctx->streams[i]->codec);
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Big, new leak

**Comment by ace20022** [comment]
@MilhouseVH sorry, should have been more precise. They work as (skin) textures, like posters, background, etc. Simplest test is to set it as background in confluence (somewhere in the skin settings).

**Comment by MilhouseVH** [comment]
Heh, seems to be working on the RPi2. I've set the Confluence "custom background" path to one of the apng's and I now have animated skiing ostriches as my backdrop...

Total CPU load is pretty reasonable, about 4-5% on a 1GHz ARM RPi2 (~20% on one core) for the [skiing ostriches](http://i.imgur.com/zinrhD2.png) apng. It has used a fairly large chunk of GPU memory to load the apng - from 240MB free to only 140MB free. The size of this file is 15MB.

[Falling snow](http://i.imgur.com/vBnzWc3.png) consumes only about 10MB GPU RAM, presumably as this is a much smaller apng (fewer frames? lower resolution?) weighing in at only 637KB.

However [HQ animated pixel art landscape](http://i.imgur.com/iXmwEJ5.png), a 17MB apng, consumes **all** available GPU memory (from 240MB free to 1MB free) yet almost 600MB of non-GPU RAM remains available.  Unsurprisingly Kodi becomes unusable with this png as a background.

Animated pngs seem to be working, but arelikely to cause memory issues on systems with limited GPU memory. Is there any reason why GPU memory is being used in favour of non-GPU RAM - maybe it's loading as a texture (or something).

**Comment by mgonzales71** [comment]
Sorry for my ignorance with the workings of Jenkins but from what I can gather from http://jenkins.kodi.tv/job/BuildMulti-PR/4321/ the Windows test build of the PR failed?  Well marked "unstable" actually... but regardless - not sure of where to download a test binary other then mirrors.kodi.tv /nightlies or /test-builds when one is available to download and test.  Apologies if I missed an obvious download location for the Kodi Win Setup that the PR test build console output says is located at c:\jenkins\slave\workspace\WIN-32\project\Win32BuildSetup\KodiSetup-20160128-df84916-HEAD.exe :grinning:  

Thanks :)

**Comment by afedchin** [comment]
@mgonzales71 here you go http://mirrors.kodi.tv/test-builds/win32/

**Comment by fritsch** [comment]
Nearly :-) Currently the whole mirror is down.

2016-01-29 7:10 GMT+01:00 Anton Fedchin <notifications@github.com>:

> @mgonzales71 <https://github.com/mgonzales71> here you go
> http://mirrors.kodi.tv/test-builds/win32/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8937#issuecomment-176593890>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by afedchin** [comment]
@fritsch it works for me.

**Comment by fritsch** [comment]
Nice - @kib also fixed it for the whole world some minutes ago.

2016-01-29 7:55 GMT+01:00 Anton Fedchin <notifications@github.com>:

> @fritsch <https://github.com/fritsch> it works for me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8937#issuecomment-176606309>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by ace20022** [comment]
@mgonzales71 http://mirrors.kodi.tv/test-builds/win32/KodiSetup-20160128-134a3dc-ffmpeg_apng.exe
Some addons failed to build: http://jenkins.kodi.tv/view/Windows/job/WIN-32/7626/

**Comment by ace20022** [comment]
@MilhouseVH that one https://i.imgur.com/q2SSe9S.png (probably the biggest one) needs almost 3gb ram for decoding, after some seconds it's only 16 mb or so for the kodi process, overall is still high. I have no clue how windows handles that. The gpu mem is almost completely used.

This apng has 91 frames á 3840x2160 pixels, i.e. 91*33mb = 3003 mb.

The only thing I could do is to add an upper bound for the memory to be used and limit the number of frames accordingly. Until I'm in the mood to implement on demand frame extraction/caching.
@fritsch @MilhouseVH @popcornmix  thoughts?

**Comment by popcornmix** [comment]
It seems quite easy for an animated png to exhaust texture memory and potentially kill kodi (if it doesn't crash immediately any subsequent gui elements will fail to allocate and so not be visible making kodi pretty much unusable).
So some sort of sanity check on size seems useful. But I don't see an easy way of choosing a size that will be optimal across many devices with different amounts of texture memory.

On demand extraction sounds like a better solution. q2SSe9S.png is only 10MB as a png, but seems to expand 30 times when decoded. On-demand extraction may increase CPU usage significantly, but that still seems like a better solution (at least for files larger than some threshold). High cpu seems better than exhausted memory and unusable kodi.

**Comment by mgonzales71** [comment]
@ace20022 Thanks for the link - also the png ( https://i.imgur.com/q2SSe9S.png ) brings Firefox to it knees even on Windows but like others have noted apngs seems to eat up texture mem quickly.

For what it's worth the use I see for apng's in Kodi are obviously for skin's - add-on icons, clearart, clearlogo, character art, home tile/menu graphics and various skin elements (progress, buttons, weather icon, fanart, controls, etc)  I would expect these uses to be small to modest in size but could be numerous. The other use would be backgrounds and that is where things could get ugly - depends on the apng.. badly created/pooly optimized images being the most likely flame for that fire.

silly question as I am sure that all the options were looked into but is ffmpeg's the best implementation of apng support for Kodi?

I will keep testing and provide feedback of any problems I run into on Windows.. thanks for finally adding apng support - kudos to all the devs for all your hard works.  it's apprecited very much.

**Comment by fritsch** [comment]
> silly question as I am sure that all the options were looked into but is ffmpeg's the best implementation of apng support for Kodi?

Do you know a better solution that works on all platforms we run? Adding no overhead to the code we maintain? If you find one, let us know.

**Comment by mgonzales71** [comment]
@fritsch meant no disrespect I hope that wasn't how that Q came across. Apologies if I did.  I have nothing but praise for all your hard work.

My go to link for APNG info has been http://littlesvr.ca/apng/ and that has been mostly from a user standpoint - they do have a dev section but I am not qualified to determine suitable criteria for potential use in Kodi.

**Comment by fritsch** [comment]
The problem is not ffmpeg, but the raw decoded buffers we keep in memory at once, when decoding the whole apng.

**Comment by zag2me** [comment]
APNG's(like animated gifs) are absolutely not designed for large backgrounds or textures. Those kind of effects should be done using the video player just like machines like the PS3 do.

Saying that, it would be nice for it to fail gracefully.

**Comment by ace20022** [comment]
@popcornmix are you okay, with a limit of 12 full hd frames, approx. 8 mb each, till I have the motivation to implement the on demand loading? btw currently I have limited gifs to 10 full hd frames.

**Comment by zag2me** [comment]
To prevent confusion in things like the animated thumbnail posters thread, could it be made equal to gifs?

**Comment by ace20022** [comment]
Added a limit of 12 full hd frames, memory-wise.

Any objections?

jenkins build this please

**Comment by ace20022** [comment]
standard win32 addon failure. I'll merge it tonight.

**Comment by fritsch** [comment]
Jep, go for it. Btw. we can sit together afterwards and see how we can adjust the work to the upcoming ffmpeg API, which deprecates a whole lot of methods we actively use.  Some easy to replace others a bit more problematic.

**Comment by stefansaraev** [comment]
maybe stupid question. giflib is still (conditionaly) used in imagefactory. is this supposed to be cleaned up ? (I'll do if you say so @ace20022)

**Comment by ace20022** [comment]
Thx, I already had that, https://github.com/ace20022/xbmc/commits/remove_giflib . It' s still needed by the texture packer...

**Comment by stefansaraev** [comment]
yep, texturepacker needs giflib-native. dont bother picking if it breaks build, I'll take care of it later.

**Pull request merged**